[WIP] Add docs for binary caching troubleshooting (#132)
microsoft/vcpkg-docs: vcpkg/users/binarycaching-troubleshooting.md

---
title: Binary Caching Troubleshooting Guide
description: Troubleshooting your binary caching setup
ms.date: 8/16/2023

> ---

Reviewer: We should add the other attribute, i.e., ms.prod, ms.author, ms.topic, etc.

# Binary Caching Troubleshooting

This guide assumes a basic understanding of binary caching. To set up binary caching instead, see [here](./users/binarycaching.md).

## Print debug output

Set up vcpkg to print debugging information:

* Classic Mode: add `--debug` to your command invocation.

> * CMake Toolchain: set `-DVCPKG_INSTALL_OPTIONS="--debug"` in your CMake call.

Reviewer 2: ```suggestion
* CMake Toolchain: set `-DVCPKG_INSTALL_OPTIONS="--debug"` in your `cmake` command call or in your `CMakePresets.json` file.
```

I think we should also start recommending setting these variables in `CMakePresets.json`.

* MSBuild/Visual Studio: set the property `VcpkgAdditionalInstallOptions` to `--debug`

> ## Configuring your Binary Cache

Reviewer: Sometimes you have "Binary Cache" capitalized but other times you don't. I'm not opinionated on whether it should be capitalized, but they should be consistent.

> Consider the case of using a NuGet feed as a binary cache.

Reviewer: Should we set up zones for the various binary caching providers?

[DEBUG]   vcpkg_from_github|b743742296a114ea1b18ae99672e02f142c4eb2bef7f57d36c038bedbfb0502f
[DEBUG]   vcpkg_replace_string|d43c8699ce27e25d47367c970d1c546f6bc36b6df8fb0be0c3986eb5830bd4f1
[DEBUG] </abientries>
```

You might notice the `triplet_abi` entry has three hashes: the hash of the file content of the `x86-windows` triplet, the `windows.cmake` toolchain, and the compiler hash. Theses hashes would change if you decide to target a different platform.

### The hashes of the files used to build your library (`portfile.cmake`, `*.patch`, or `vcpkg.json`) change between runs

This usually means the baseline of your libraries is not fixed. Frequently, the problem occurs when you clone the latest repository from vcpkg, which will pull in the updated baseline of libraries.
Possible solutions:

* (recommended) The copy of the vcpkg catalog must be pinned to a specific commit. When you run `git clone vcpkg`, make sure you run `git checkout [commit_id]` or add vcpkg as a submodule in your project. The reason is the ABI of a library is calculated based on the hashes of scripts and ports in the vcpkg repository. Pulling the latest changes will likely cause frequent rebuilds.

> * Set `builtin-baseline` in your [vcpkg manifest](./users/manifests.md).

Reviewer 3: ```suggestion
* Set `builtin-baseline` in your [vcpkg manifest](./reference/vcpkg-json.md#builtin-baseline).
```

Reviewer 3: I think this reference article is a better one to use here. I am not sure how to make anchor links work with Markdown though, someone should double check what I suggested above.

> ### The hash of the vcpkg internal CMake functions (vcpkg_from_github, vcpkg_copy_pdbs, vcpkg_fixup_pkgconfig, etc) changes between runs

Reviewer 3: ```suggestion
### ABI hash of vcpkg CMake helper functions changes between runs
```

> You are always pulling the latest internal CMake functions, which are used in the ABI calculation. This is probably also caused by cloning the latest vcpkg repo. We expect this issue to go away in the future when we version these functions.

Reviewer 3: ```suggestion
Internal vcpkg helper functions like vcpkg_from_github, vcpkg_copy_pdbs, vcpkg_fixup_pkgconfig, etc. are used in the ABI calculation. If you are getting new ABI hashes for these ports, this is probably also caused by pulling the latest vcpkg repo. This issue will be addressed in a future update when these functions will be separately versioned.
```

Possible solutions:

> * See the previous section on how to pin the vcpkg catalog to a specific commit.

Reviewer 3: I think it's okay to repeat the instructions here. The idea with an article like this is that the user scrolls through it and pauses on a heading that is applicable to them. They shouldn't have to scroll somewhere else to find the specific nugget of information they need.

> ### The version of CMake, Powershell, compiler version changes between runs

Reviewer 3: ```suggestion
### Version of CMake, PowerShell, or your compiler changes between runs
```

> Somehow, your environment changes between runs. To verify that this indeed is the issue, temporarily disable including the compiler hash in your ABI by setting `VCPKG_DISABLE_COMPILER_TRACKING` to `ON` in your custom triplet. If the issue persists, fix tool versions with `--x-abi-tools-use-exact-versions`. Note it is not recommended to use these options in development or a production environment, as they can cause very hard to diagnose errors.

Reviewer 3: ```suggestion
Somehow, your environment changes between runs. To verify that this indeed is the issue, temporarily disable including the compiler hash in your ABI by setting `VCPKG_DISABLE_COMPILER_TRACKING` to `ON` in your custom triplet. If the issue persists, fix tool versions with `--x-abi-tools-use-exact-versions`. It is not recommended to use these options in development or a production environment as they can cause errors that are very hard to diagnose.
```

> In these cases, vcpkg is behaving correctly, we expect the ABI hashes to change if your toolchain changes. Rebuilds can be expensive, however, so we have recommendations for the following environments:

Reviewer 3: ```suggestion
In these cases, vcpkg is behaving correctly. It is expected for toolchain changes to cause ABI hash changes as well. Rebuilds can be expensive however, so we have some suggestions for the following environments:
```

> * CI: We’ve seen this as an issue if you are using CI (such as GitHub Hosted Runners) and using an image that may update your C++/C compiler or CMake version from underneath you. To minimize this, choose a versioned stable image if possible.
> * Visual Studio C++ Compiler Updates: Major, minor and even patch updates to the compiler will cause your library ABI to change.  To minimize this, stop automatic Visual Studio updates.

Reviewer 3: ```suggestion
* Continuous integration systems: Some continuous integration systems (such as GitHub Hosted Runners) use images that can update your C/C++ compiler or CMake version periodically. To minimize this, choose a versioned stable image if possible.
* MSVC compiler updates: Major, minor and patch updates to the MSVC compiler will cause your library ABI hash to change. To minimize this, turn off automatic Visual Studio updates.
```

If the above options do not work, consider the following solutions:

> * Use `vcpkg export` to package your libraries instead (currently, only works in classic mode)

Reviewer 3: Isn't the export for manifest mode feature available now? Worth double checking.

* Consider using a Docker image to build your libraries

> * Have an auxiliary CI run that builds vcpkg libraries daily or weekly

Reviewer 3: ```suggestion
* Have an auxiliary continuous integration run that builds vcpkg libraries on a regular cadence (e.g. daily or weekly)
```
